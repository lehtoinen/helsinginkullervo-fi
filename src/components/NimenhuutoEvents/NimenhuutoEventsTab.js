import React from 'react';
import PropTypes from 'prop-types';
import CSV from 'comma-separated-values';
import moment from 'moment';

import styles from './NimenhuutoEventsTab.module.scss';

const NimenhuutoEventsTab = ({ label, items, loading, error }) => {
  let state;
  if (error) {
    return <div className="error">Lataaminen epäonnistui!</div>;
  } else if (loading && !items.length) {
    return <div className="loading">Ladataan...</div>;
  } else if (!loading && !items.length) {
    return <div className="empty">Ei tulevia tapahtumia.</div>;
  }

  return (
    <div className={styles.root}>
      {items.map(item => {
        const { subject, location, date, time, url, description } = item;
        return (
          <a
            href={url}
            target="_blank"
            rel="noopener noreferrer"
            className={`event-row ${styles.row}`}
            key={`${subject}${date}${time}`}
          >
            <DateBadge dateStr={date} />
            <div className={styles.info}>
              <div className={styles.subject}>
                {/* subject contains the name of the nimenhuuto-instance (for example "Kullervo: ...")
                stripping that to avoid unnecessary repetition */
                subject.substr(subject.indexOf(': ') + 2)}
                {/* {description && <span>&nbsp;i</span>} */}
              </div>
              <div className={styles.details}>
                <span className={styles.time}>Klo {time}</span>
                <span className={styles.location}>{location}</span>
              </div>
            </div>
          </a>
        );
      })}
    </div>
  );
};

NimenhuutoEventsTab.propTypes = {
  label: PropTypes.string.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.bool.isRequired,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      subject: PropTypes.string.isRequired,
      location: PropTypes.string.isRequired,
      date: PropTypes.date,
      time: PropTypes.string.isRequired,
      url: PropTypes.string.isRequired,
      description: PropTypes.string,
    })
  ).isRequired,
};

const DateBadge = ({ dateStr }) => {
  moment.locale('fi');
  const eventDate = moment(dateStr);
  const weekday = eventDate.format('dd');
  const date = eventDate.format('DD.MM');

  return (
    <div className={styles.day}>
      <div className={styles.weekday}>{weekday}</div>
      <div className={styles.date}>{date}</div>
    </div>
  );
};

export default NimenhuutoEventsTab;

// TODO/IMPROVEMENT: Add a collapse for the description when available  (> / ▼ -icon)
