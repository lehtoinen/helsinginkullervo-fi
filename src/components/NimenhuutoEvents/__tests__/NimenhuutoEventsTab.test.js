import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import NimenhuutoEventsTab from '../NimenhuutoEventsTab';

configure({ adapter: new Adapter() });

let baseProps = {
  label: 'Foo',
  loading: false,
  error: false,
  items: [
    {
      subject: 'Foo',
      location: 'Foo',
      date: new Date('2018-03-23'),
      time: 'Foo',
      url: 'Foo',
    },
  ],
};

describe('NimenhuutoEventsTab', () => {
  test('renders all items', () => {
    const props = { ...baseProps };
    const component = mount(<NimenhuutoEventsTab {...props} />);

    expect(component.find('.event-row')).toHaveLength(baseProps.items.length);
    expect(component).toMatchSnapshot();
  });

  test('renders "error" when { error: true }', () => {
    const props = { ...baseProps, ...{ error: true } };
    const component = mount(<NimenhuutoEventsTab {...props} />);

    expect(component.find('.error')).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  test('renders "loading" when { loading: true, items: [] }', () => {
    const props = { ...baseProps, ...{ loading: true, items: [] } };
    const component = mount(<NimenhuutoEventsTab {...props} />);

    expect(component.find('.loading')).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  test('renders "empty" when { loading: false, items: [] }', () => {
    const props = { ...baseProps, ...{ loading: false, items: [] } };
    const component = mount(<NimenhuutoEventsTab {...props} />);

    expect(component.find('.empty')).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });
});
