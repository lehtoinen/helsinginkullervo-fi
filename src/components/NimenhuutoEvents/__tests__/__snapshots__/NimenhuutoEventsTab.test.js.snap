// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NimenhuutoEventsTab renders "empty" when { loading: false, items: [] } 1`] = `
<NimenhuutoEventsTab
  error={false}
  items={Array []}
  label="Foo"
  loading={false}
>
  <div
    className="empty"
  >
    Ei tulevia tapahtumia.
  </div>
</NimenhuutoEventsTab>
`;

exports[`NimenhuutoEventsTab renders "error" when { error: true } 1`] = `
<NimenhuutoEventsTab
  error={true}
  items={
    Array [
      Object {
        "date": 2018-03-23T00:00:00.000Z,
        "location": "Foo",
        "subject": "Foo",
        "time": "Foo",
        "url": "Foo",
      },
    ]
  }
  label="Foo"
  loading={false}
>
  <div
    className="error"
  >
    Lataaminen ep√§onnistui!
  </div>
</NimenhuutoEventsTab>
`;

exports[`NimenhuutoEventsTab renders "loading" when { loading: true, items: [] } 1`] = `
<NimenhuutoEventsTab
  error={false}
  items={Array []}
  label="Foo"
  loading={true}
>
  <div
    className="loading"
  >
    Ladataan...
  </div>
</NimenhuutoEventsTab>
`;

exports[`NimenhuutoEventsTab renders all items 1`] = `
<NimenhuutoEventsTab
  error={false}
  items={
    Array [
      Object {
        "date": 2018-03-23T00:00:00.000Z,
        "location": "Foo",
        "subject": "Foo",
        "time": "Foo",
        "url": "Foo",
      },
    ]
  }
  label="Foo"
  loading={false}
>
  <div
    className="root"
  >
    <a
      className="event-row row"
      href="Foo"
      key="FooFri Mar 23 2018 02:00:00 GMT+0200 (EET)Foo"
      rel="noopener noreferrer"
      target="_blank"
    >
      <DateBadge
        dateStr={2018-03-23T00:00:00.000Z}
      >
        <div
          className="day"
        >
          <div
            className="weekday"
          >
            pe
          </div>
          <div
            className="date"
          >
            23.03
          </div>
        </div>
      </DateBadge>
      <div
        className="info"
      >
        <div
          className="subject"
        >
          oo
        </div>
        <div
          className="details"
        >
          <span
            className="time"
          >
            Klo 
            Foo
          </span>
          <span
            className="location"
          >
            Foo
          </span>
        </div>
      </div>
    </a>
  </div>
</NimenhuutoEventsTab>
`;
